server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java

```diff
@@ -725,19 +719,30 @@ private TSExecuteStatementResp internalExecuteQueryStatement(String statement,
       if (plan instanceof AuthorPlan) {
         plan.setLoginUserName(username);
       }
+
+      TSExecuteStatementResp resp = null;
+
+      if (plan instanceof QueryPlan) {
+        resp = getQueryColumnHeaders(plan, username);
+      }
```

Comment:
```suggestion
      // execute it before createDataSet since it may change the content of query plan
      if (plan instanceof QueryPlan) {
        resp = getQueryColumnHeaders(plan, username);
      }
```